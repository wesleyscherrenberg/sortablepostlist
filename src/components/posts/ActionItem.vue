<script setup lang="ts">
import {computed} from "vue";
import type {ActionItemProp, ActionsType} from "@/types/Types";

const emit = defineEmits(['onTimeTravelChange'])

const props = defineProps<ActionItemProp>()

const onTimeTravelHandler = (action: ActionsType) => {
  emit('onTimeTravelChange', action)
}

const itemText = computed(() => {
  return `Moved post ${props.action.postId} from index ${props.action.fromIndex} to index ${props.action.toIndex}`
})
</script>

<template>
  <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
    <p class="flex-1" data-test="itemText">{{itemText}}</p>
    <button @click="onTimeTravelHandler(action)"
            aria-label="onTimeTravelBtn"
            role="button"
            data-test="onTimeTravelBtn"
            class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded ">
      {{btnText}}
    </button>
  </div>
</template>
