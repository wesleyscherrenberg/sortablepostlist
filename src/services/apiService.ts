import type {PostsType} from "@/types/Types";

export async function fetchPosts(): Promise<PostsType[]> {

    const response = await fetch('https://jsonplaceholder.typicode.com/posts');

    if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
    }

    return response.json();
}
