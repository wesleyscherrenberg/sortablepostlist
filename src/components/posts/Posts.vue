<script setup lang="ts">
import {onBeforeMount, ref} from "vue";
import {fetchPosts} from "@/services/apiService.js";
import type {ActionsType, PostsPropType, PostsType} from "@/types/Types";

const INFO = 'info';
const ERROR = 'error';

const props = defineProps<PostsPropType>()

const actions = ref<ActionsType[]>([])
const posts = ref<PostsType[]>([])
const toastMessage = ref('');
const toastVariant = ref(INFO);

const fetchData = async () => {
  try {
    const data = await fetchPosts();
    posts.value = props.limit ? data.slice(0, props.limit) : data;

    const postLabel = posts.value.length === 1 ? 'post' : 'posts'

    toastMessage.value = `Hello! I have provided ${posts.value.length} ${postLabel} for you to sort.`;
    toastVariant.value = INFO;

  } catch (err) {
    let message = 'There was an error loading the posts. Please try again later.';

    if(err instanceof Error) {
      message = err.message
    }
    toastMessage.value = message;
    toastVariant.value = ERROR;

  }
}

const sortPosts = (postId: number, toIndex: number) => {
  const _posts = [...posts.value]
  const postIndex = _posts.findIndex((post) => post.id === postId)
  const currentPost = _posts.splice(postIndex, 1)[0]
  _posts.splice(toIndex, 0, currentPost)
  posts.value = _posts
}

const onSortChange = (action: ActionsType) => {
  actions.value.unshift(action)
  sortPosts(action.postId, action.toIndex)
}

const onTimeTravelChange = (action: ActionsType) => {
  const actionIndex = actions.value.indexOf(action);
  if (actionIndex < 0) return;

  actions.value.slice(0, actionIndex + 1)
      .forEach(action => sortPosts(action.postId, action.fromIndex));

  actions.value = actions.value.slice(actionIndex + 1);

}

onBeforeMount(fetchData);

</script>

<template>
  <Post :title="'Sortable Post List'">
    <transition-group v-if="posts.length" class="flex flex-col gap-4" name="list" tag="div">
      <PostItem
          v-for="(post, index) in posts"
          :key="post.id"
          :index="index"
          :postId="post.id"
          :showSortUp="index > 0"
          :showSortDown="index < posts.length - 1"
          @onSortChange="onSortChange"
      />
    </transition-group>
    <Notification v-else
                  icon="sad"
                  :text="'There a no post available at this moment'"/>
  </Post>

  <Card :title="'List of actions commited'">
    <ActionList v-if="actions.length">
      <ActionItem
          v-for="(action, index) in actions"
          :key="index"
          :action="action"
          :btn-text="'Time travel'"
          @onTimeTravelChange="onTimeTravelChange"
      />
    </ActionList>
    <Notification v-else
                  icon="info"
                  :text="'Please sort a post in the Sortable Post List'"/>
  </Card>

  <Toast :message="toastMessage" :variant="toastVariant" />
</template>
