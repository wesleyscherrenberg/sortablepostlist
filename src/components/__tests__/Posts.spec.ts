import {shallowMount, VueWrapper} from '@vue/test-utils'
import {fetchPosts} from "@/services/apiService.js";
import {beforeAll, beforeEach, describe, expect, it, vi} from "vitest";
import Posts from "@/components/posts/Posts.vue";
import ActionItem from "@/components/posts/ActionItem.vue";
import type {ActionsType} from "@/types/Types";
import {ActionListStub, CardStub, NotificationStub, PostItemStub, PostStub, ToastStub} from "./stubs";

const mocks = vi.hoisted(() => {
    return {
        fetchPosts: vi.fn().mockResolvedValue(
            Array.from({length: 100}, (_, i) => ({
                userId: i + 1,
                id: i + 1,
                title: "title",
                body: "body"
            }))
        ),
    }
})

vi.mock('@/services/apiService', () => ({
    fetchPosts: mocks.fetchPosts,
}));

const factory = () => {
    return shallowMount(Posts, {
        props: {
            limit: 5
        },
        global: {
            stubs: {
                Post: PostStub,
                Card: CardStub,
                PostItem: PostItemStub,
                ActionItem: ActionItem,
                ActionList: ActionListStub,
                Toast: ToastStub,
                Notification: NotificationStub

            }
        }
    });
}

const sortCases = [
    {
        postId: 2,
        fromIndex: 1,
        toIndex: 2
    },
    {
        postId: 4,
        fromIndex: 3,
        toIndex: 4
    },
    {
        postId: 3,
        fromIndex: 1,
        toIndex: 0
    }
]

describe('Posts.vue', () => {

    let wrapper: VueWrapper;

    beforeAll( () => {
        wrapper = factory();
    });

    describe('When the component renders', () => {

        it('matches the snapshot', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        it('should have postItems', () => {
            expect(fetchPosts).toHaveBeenCalled()

            expect(wrapper.findAllComponents(PostItemStub).length).toBe(5);
        })

        it('Toast message and variant attributes should have data', () => {

            expect(wrapper.findComponent(ToastStub).attributes('message'))
                .toEqual(`Hello! I have provided 5 posts for you to sort.`);

            expect(wrapper.findComponent(ToastStub).attributes('variant'))
                .toEqual('info');
        })

    })

    describe.each(sortCases)('when sorting post List', (action: ActionsType) => {

        const { fromIndex, toIndex, postId} = action
        const sortCaseIndex = sortCases.indexOf(action);

        it(`Move post ${postId} from index ${fromIndex} to index ${toIndex}`, async () => {

            const postItem = wrapper.findAllComponents(PostItemStub)[fromIndex];
            await postItem.vm.$emit('onSortChange', action);

            expect(wrapper.findAllComponents(PostItemStub)[toIndex].attributes('postid'))
                .toEqual(postId.toString());

        })

        it(`Action item should have text Moved post ${postId} from index ${fromIndex} to index ${toIndex} action`, () => {
            const reverseIndex = wrapper.findAllComponents(ActionItem).length - sortCaseIndex - 1;
            const actionItemComponent = wrapper.findAllComponents(ActionItem)[reverseIndex];
            expect(actionItemComponent.find('[data-test="itemText"]').text())
                .toEqual(`Moved post ${postId} from index ${fromIndex} to index ${toIndex}`);
        })

        it(`There should be ${sortCaseIndex + 1} list of action items committed`, () => {
            expect(wrapper.findAllComponents(ActionItem).length).toBe(sortCaseIndex + 1);
        })

        it('matches the snapshot', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

    })

    describe('When Time Travel on the second action card', () => {

        beforeAll(async () => {
            const actionItemComponent = wrapper.findAllComponents(ActionItem)[1]
            await actionItemComponent.find('[data-test="onTimeTravelBtn"]').trigger('click')
        })

        it('should only have 1 action item left', () => {
            expect(wrapper.findAllComponents(ActionItem)).toHaveLength(1)
        })

        it('Action item should only have text Moved post 2 from index 1 to index 2 action', () => {
            const actionItemComponent = wrapper.findAllComponents(ActionItem)[0];
            expect(actionItemComponent.find('[data-test="itemText"]').text()).toEqual('Moved post 2 from index 1 to index 2');
        })

        it('Post list should match Post 1,3,2,4,5', () => {

            const postItems = wrapper.findAllComponents(PostItemStub);
            const postIds = [1, 3, 2, 4, 5];

            postIds.forEach((postId, index) => {
                expect(postItems[index].attributes('postid')).toEqual(postId.toString());
            });
        })


    })

    describe('When there are no posts', () => {

        beforeEach( () => {
            mocks.fetchPosts.mockResolvedValue([])
            wrapper = factory()
        })

        it('should show the postNotification', () => {
            expect(wrapper.findComponent(NotificationStub).exists()).toBe(true);
            expect(wrapper.findComponent(NotificationStub).attributes('icon')).toEqual('sad');
            expect(wrapper.findComponent(NotificationStub).attributes('text')).toEqual('There a no post available at this moment');
        });
    })

    describe('When fails with exception', () => {

        beforeEach( () => {
            mocks.fetchPosts.mockRejectedValue(new Error('Async error'))
            wrapper = factory()
        })

        it('should show error toast when fetchPosts fails', async () => {
            expect(wrapper.findComponent(ToastStub).attributes('message')).toEqual('Async error');
            expect(wrapper.findComponent(ToastStub).attributes('variant')).toEqual('error');
        });

    })
})
