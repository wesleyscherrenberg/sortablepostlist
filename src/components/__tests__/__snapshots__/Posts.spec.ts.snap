// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Posts.vue > When the component renders > matches the snapshot 1`] = `
"<div title="Sortable Post List">
  <transition-group-stub name="list" tag="div" appear="false" persisted="false" css="true" class="flex flex-col gap-4">
    <div index="0" postid="1" showsortup="false" showsortdown="true"></div>
    <div index="1" postid="2" showsortup="true" showsortdown="true"></div>
    <div index="2" postid="3" showsortup="true" showsortdown="true"></div>
    <div index="3" postid="4" showsortup="true" showsortdown="true"></div>
    <div index="4" postid="5" showsortup="true" showsortdown="false"></div>
  </transition-group-stub>
</div>
<div title="List of actions commited">
  <div icon="info" text="Please sort a post in the Sortable Post List"></div>
</div>
<div message="Hello! I have provided 5 posts for you to sort." variant="info"></div>"
`;

exports[`Posts.vue > when sorting post List > matches the snapshot 1`] = `
"<div title="Sortable Post List">
  <transition-group-stub name="list" tag="div" appear="false" persisted="false" css="true" class="flex flex-col gap-4">
    <div index="0" postid="1" showsortup="false" showsortdown="true"></div>
    <div index="1" postid="3" showsortup="true" showsortdown="true"></div>
    <div index="2" postid="2" showsortup="true" showsortdown="true"></div>
    <div index="3" postid="4" showsortup="true" showsortdown="true"></div>
    <div index="4" postid="5" showsortup="true" showsortdown="false"></div>
  </transition-group-stub>
</div>
<div title="List of actions commited">
  <div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 2 from index 1 to index 2</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
  </div>
</div>
<div message="Hello! I have provided 5 posts for you to sort." variant="info"></div>"
`;

exports[`Posts.vue > when sorting post List > matches the snapshot 2`] = `
"<div title="Sortable Post List">
  <transition-group-stub name="list" tag="div" appear="false" persisted="false" css="true" class="flex flex-col gap-4">
    <div index="0" postid="1" showsortup="false" showsortdown="true"></div>
    <div index="1" postid="3" showsortup="true" showsortdown="true"></div>
    <div index="2" postid="2" showsortup="true" showsortdown="true"></div>
    <div index="3" postid="5" showsortup="true" showsortdown="true"></div>
    <div index="4" postid="4" showsortup="true" showsortdown="false"></div>
  </transition-group-stub>
</div>
<div title="List of actions commited">
  <div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 4 from index 3 to index 4</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 2 from index 1 to index 2</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
  </div>
</div>
<div message="Hello! I have provided 5 posts for you to sort." variant="info"></div>"
`;

exports[`Posts.vue > when sorting post List > matches the snapshot 3`] = `
"<div title="Sortable Post List">
  <transition-group-stub name="list" tag="div" appear="false" persisted="false" css="true" class="flex flex-col gap-4">
    <div index="0" postid="3" showsortup="false" showsortdown="true"></div>
    <div index="1" postid="1" showsortup="true" showsortdown="true"></div>
    <div index="2" postid="2" showsortup="true" showsortdown="true"></div>
    <div index="3" postid="5" showsortup="true" showsortdown="true"></div>
    <div index="4" postid="4" showsortup="true" showsortdown="false"></div>
  </transition-group-stub>
</div>
<div title="List of actions commited">
  <div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 3 from index 1 to index 0</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 4 from index 3 to index 4</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
    <div class="p-4 flex items-center border-b border-theme-light last:border-none transform transition-all duration-500">
      <p class="flex-1" data-test="itemText">Moved post 2 from index 1 to index 2</p><button aria-label="onTimeTravelBtn" role="button" data-test="onTimeTravelBtn" class="p-4 bg-theme-secondary py-3 px-4 flex-none rounded">Time travel</button>
    </div>
  </div>
</div>
<div message="Hello! I have provided 5 posts for you to sort." variant="info"></div>"
`;
